File: src/providers/tiptap/rules/appendTransaction-full-scan.ts
/**
 * tiptap-appendTransaction-full-scan (reliability)
 *
 * Detects `appendTransaction` callbacks that call `doc.descendants(...)` on
 * every transaction without first checking whether the document changed.
 * O(n) per keystroke is avoidable with a simple docChanged guard.
 */
import { walk } from '../utils.js';

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'appendTransaction must guard doc.descendants() with a docChanged check',
      category: 'reliability',
      rationale:
        'doc.descendants() visits every node in the document on every transaction — every keystroke, selection change, and formatting toggle. For documents with many nodes this is O(n) per transaction and causes visible input latency. ProseMirror provides the docChanged flag precisely to avoid this work when the document structure has not changed.',
      docsUrl: 'https://prosemirror.net/docs/ref/#state.PluginSpec.appendTransaction',
      recommended: true,
    },
    messages: {
      fullScanOnEveryTransaction:
        'appendTransaction calls doc.descendants() on every transaction. Check docChanged first: const docChanged = transactions.some(tr => tr.docChanged); if (!docChanged) return;',
    },
    schema: [],
  },
  create(context: any) {
    function hasDocChangedGuard(fnBody: any): boolean {
      let found = false;
      walk(fnBody, (node: any) => {
        if (found) return false;
        // Look for tr.docChanged or transactions.some(tr => tr.docChanged) or docChanged identifier used as condition
        if (
          node?.type === 'MemberExpression' &&
          node.property?.name === 'docChanged'
        ) {
          found = true;
          return false;
        }
        // Also look for if (!docChanged) return; pattern
        if (
          node?.type === 'Identifier' &&
          node.name === 'docChanged'
        ) {
          found = true;
          return false;
        }
      });
      return found;
    }

    function hasDescendantsCall(fnBody: any): boolean {
      let found = false;
      walk(fnBody, (node: any) => {
        if (found) return false;
        if (
          node?.type === 'CallExpression' &&
          node.callee?.type === 'MemberExpression' &&
          node.callee.property?.name === 'descendants'
        ) {
          found = true;
        }
      });
      return found;
    }

    function checkAppendTransactionFn(fn: any, reportNode: any): void {
      const body = fn?.body;
      if (!body) return;
      if (hasDescendantsCall(body) && !hasDocChangedGuard(body)) {
        context.report({ node: reportNode, messageId: 'fullScanOnEveryTransaction' });
      }
    }

    return {
      Property(node: any) {
        const keyName =
          node.key?.type === 'Identifier'
            ? node.key.name
            : node.key?.type === 'Literal'
            ? node.key.value
            : null;
        if (keyName !== 'appendTransaction') return;

        const val = node.value;
        if (
          val?.type === 'FunctionExpression' ||
          val?.type === 'ArrowFunctionExpression'
        ) {
          checkAppendTransactionFn(val, node);
        }
      },
    };
  },
};

export const tiptapAppendTransactionFullScanRule = rule;
export default rule;

File: src/providers/tiptap/rules/python/append_transaction_full_scan.py
"""tiptap-appendTransaction-full-scan (python analog, reliability)

Same rationale as the JS rule, ported for Python code that reimplements a
ProseMirror-style `appendTransaction` hook (custom collaboration/sync
backends): flags a `def append_transaction(...)` whose body calls
`doc.descendants(...)` on every invocation without first checking a
`doc_changed`/`docChanged` guard.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))

from utils import APPEND_TRANSACTION_NAMES, func_or_method_name, function_named, loc  # noqa: E402

RULE_KEY = "tiptap-appendTransaction-full-scan"

_DOC_CHANGED_NAMES = frozenset({"docChanged", "doc_changed"})


def _has_doc_changed_guard(fn: ast.AST) -> bool:
    for node in ast.walk(fn):
        if isinstance(node, ast.Attribute) and node.attr in _DOC_CHANGED_NAMES:
            return True
        if isinstance(node, ast.Name) and node.id in _DOC_CHANGED_NAMES:
            return True
    return False


def _has_descendants_call(fn: ast.AST) -> bool:
    for node in ast.walk(fn):
        if isinstance(node, ast.Call) and func_or_method_name(node.func) == "descendants":
            return True
    return False


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []

    for node in ast.walk(tree):
        if not function_named(node, *APPEND_TRANSACTION_NAMES):
            continue
        if _has_descendants_call(node) and not _has_doc_changed_guard(node):
            line, col, end_line, end_col = loc(node)
            out.append(
                {
                    "file": path,
                    "line": line,
                    "column": col,
                    "endLine": end_line,
                    "endColumn": end_col,
                    "ruleKey": RULE_KEY,
                }
            )

    return out
